### PR TITLE
storage: do not gather up too many intents/keys

### PR DESCRIPTION
**Pull request by tbg**
Fallout from #18199 and corresponding testing in #15997. I think it'll be nontrivial to max out
these budgets in practice, but I can definitely do it in intentionally evil tests, and it's good to
know that there is some rudimentary form of memory accounting in this queue.

### CODE REVIEW

File: pkg/storage/gc_queue.go [hunk 1]
```diff
@@ -767,30 +774,50 @@ func RunGC(
...
 					}
 					// With an active intent, GC ignores MVCC metadata & intent value.
 					startIdx = 2
 				}
+				if keyBytes > keyBudgetBytes {
+					// Stop considering keys for GC if we've already crashed the budget.
+					//
+					// TODO(tschottdorf): we already send multiple GCRequests if we have too many keys,
+					// but we should be dispatching them as we scan to avoid holding the whole chunk of
+					// keys in memory. Large Ranges should be the exception, but they can definitely be
```
Comment (nvanbenschoten) [inline pkg/storage/gc_queue.go]: nit: `s/Ranges/ranges/`

File: pkg/storage/gc_queue.go [hunk 2]
```diff
@@ -751,6 +751,13 @@ func RunGC(
 	txnMap := map[uuid.UUID]*roachpb.Transaction{}
 	intentSpanMap := map[uuid.UUID][]roachpb.Span{}
 
+	const (
+		intentBudgetBytes = 30 * 1 << 20 // 30mb
```
Comment (nvanbenschoten) [inline pkg/storage/gc_queue.go]: `intentBytesBudget` and `keyBytesBudget` sound better to me. Take it or leave it.

File: pkg/storage/gc_queue.go [hunk 3]
```diff
@@ -767,30 +774,50 @@ func RunGC(
...
-					// multiple requests in the event that more than X keys
-					// are added to the request.
 					gcKeys = append(gcKeys, roachpb.GCRequest_GCKey{Key: expBaseKey, Timestamp: gcTS})
+					keyBytes += len(expBaseKey)
 				}
 			}
 		}
 	}
 
+	if intentBytes > intentBudgetBytes {
```
Comment (nvanbenschoten) [inline pkg/storage/gc_queue.go]: Will this... work? We haven't called `processKeysAndValues` yet.

File: pkg/storage/gc_queue.go [hunk 4]
```diff
@@ -767,30 +774,50 @@ func RunGC(
...
 				if meta.Txn != nil {
 					// Keep track of intent to resolve if older than the intent
-					// expiration threshold.
-					if meta.Timestamp.Less(intentExp) {
+					// expiration threshold. Don't do this if we've already collected
+					// a lot of potential intents (note that what we collect here is
+					// extremely likely to be collectable unless the user is running
+					// very large numbers of very long transactions). But the same
+					// consideration as the TODO below holds: we should be handling
+					// the chunk and resetting the budget.
```
Comment (nvanbenschoten) [inline pkg/storage/gc_queue.go]: `s/handling the chunks/handling the chunks as we scan`

File: pkg/storage/gc_queue.go (hunk 3 above)
Comment (bdarnell) [reply pkg/storage/gc_queue.go]: Yeah, I think this needs a test.

**Comment by tschottdorf** [comment]
Obviated by @spencerkimball's improvements.

**Pull request closed without merging**